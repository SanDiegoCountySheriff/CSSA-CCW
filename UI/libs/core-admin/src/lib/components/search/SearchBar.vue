<!-- eslint-disable vue/singleline-html-element-content-newline -->
<template>
  <div>
    <v-autocomplete
      v-model="state.model"
      :items="items"
      :color="$vuetify.theme.dark ? 'white' : 'primary'"
      :loading="state.isLoading"
      :search-input.sync="search"
      item-text="Name"
      item-value="orderId"
      :filter="filterData"
      placeholder="Start typing to search"
      prepend-icon="mdi-magnify"
      no-data-text="No results found..."
      clearable
      append-icon=""
      return-object
    >
      <template #item="{ item }">
        <router-link
          :to="{
            name: 'PermitDetail',
            params: { orderId: item.orderId },
          }"
          tag="a"
          target="_self"
          style="text-decoration: none; color: inherit"
          replace
        >
          <v-row>
            <v-col>
              {{ item.Name }}, {{ item.dob.birthDate }},
              <v-chip
                medium
                label
                :color="
                  $vuetify.theme.dark ? '' : item.isComplete ? 'blue' : 'error'
                "
                class="ml-4 white--text"
              >
                {{ item.isComplete ? 'Ready for review' : 'Incomplete' }}
              </v-chip>

              <v-chip
                medium
                label
                color="accent lighten-2"
                class="ml-4"
              >
                {{ item.orderId }}
              </v-chip>
            </v-col>
          </v-row>
        </router-link>
      </template>
      <template #selection=""> </template>
    </v-autocomplete>
  </div>
</template>
<script setup lang="ts">
import { usePermitsStore } from '@core-admin/stores/permitsStore'
import { computed, reactive, ref, watch } from 'vue'

const permitStore = usePermitsStore()

const state = reactive({
  entries: [],
  isLoading: false,
  model: null,
})

const search = ref('')

const items = computed(() => {
  return state.entries.map(entry => {
    const Name = entry?.firstName || ''

    return { ...entry, Name }
  })
})

function filterData(item, queryText) {
  return (
    item.dob.birthDate.toLowerCase().includes(queryText.toLowerCase()) ||
    item.currentAddress.streetAddress
      .toLowerCase()
      .includes(queryText.toLowerCase()) ||
    item.firstName.toLowerCase().includes(queryText.toLowerCase()) ||
    item.lastName.toLowerCase().includes(queryText.toLowerCase()) ||
    item.email.toLowerCase().includes(queryText.toLowerCase()) ||
    item.address.zip.toLowerCase().includes(queryText.toLowerCase())
  )
}

watch(search, async () => {
  // Items have already been loaded
  if (search.value && search.value.length > 2) {
    // Items have already been requested
    if (state.isLoading) return

    state.isLoading = true

    // Lazily load input items
    await permitStore.getPermitSearchApi(search.value)
    state.entries = permitStore.getSearchResults

    state.isLoading = false
  } else {
    permitStore.setSearchResults([])
    state.entries = []
  }
})
</script>

<style lang="scss" scoped>
.v-menu__content {
  z-index: 99999;
}
.v-list-item {
  position: absolute;
  bottom: 20px;

  &__icon {
    margin: 0;
  }

  &__content {
    padding: 0;
    margin-left: 9px;
  }

  &__title {
    text-align: left;
  }

  &__subtitle {
    text-align: left;
    text-overflow: ellipsis;
    font-size: 0.76rem;
  }
}
</style>
