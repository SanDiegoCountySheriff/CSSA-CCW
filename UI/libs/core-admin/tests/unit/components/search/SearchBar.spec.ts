/* eslint-disable @typescript-eslint/ban-ts-comment */
import SearchBar from '@core-admin/components/search/SearchBar.vue'
import Vuetify from 'vuetify'
import { createLocalVue, mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'

const localVue = createLocalVue()
const pinia = createTestingPinia()
const tMock = {
  $t: text => text,
}

describe('SearchBar', () => {
  let vuetify
  let wrapper

  beforeEach(() => {
    vuetify = new Vuetify()
    //@ts-ignore
    wrapper = mount(SearchBar, {
      localVue,
      vuetify,
      pinia,
      mocks: tMock,
    })
  })
  afterEach(() => {
    wrapper.destroy()
  })

  it('should match the snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
