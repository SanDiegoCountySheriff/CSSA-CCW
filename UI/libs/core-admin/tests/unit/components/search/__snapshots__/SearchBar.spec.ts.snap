// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match the snapshot 1`] = `
<div>
  <div class="v-input search theme--light v-text-field v-text-field--placeholder v-select v-autocomplete">
    <div class="v-input__prepend-outer">
      <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--light"></i></div>
    </div>
    <div class="v-input__control">
      <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-2" class="v-input__slot">
        <div class="v-select__slot">
          <div class="v-select__selections"><input id="input-2" placeholder="Start typing to search" type="text" autocomplete="off"></div>
          <div class="v-input__append-inner">
            <div></div>
          </div><input type="hidden">
        </div>
        <div class="v-menu">
          <!---->
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
        </div>
      </div>
    </div>
  </div>
</div>
`;
